deep_qa/data/instance.py

```diff
@@ -0,0 +1,62 @@
+from typing import Dict, List
+
+import numpy
+
+from .fields import Field
+from . import Vocabulary
+
+
+class Instance:
+    """
+    An ``Instance`` is a collection of :class:`Field` objects, specifying the inputs and outputs to
+    some model.  We don't make a distinction between inputs and outputs here, though - all
+    operations are done on all fields, and when we return arrays, we return them as dictionaries
+    keyed by field name.  A model can then decide which fields it wants to use as inputs as which
+    as outputs.
+
+    The ``Fields`` in an ``Instance`` can start out either indexed or un-indexed.  During the data
+    processing pipeline, all fields will end up as ``IndexedFields``, and will then be converted
+    into padded arrays by a ``DataGenerator``.
+    """
+    def __init__(self, fields: Dict[str, Field]):
+        self._fields = fields
+
+    def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
+        """
+        Increments counts in the given ``counter`` for all of the vocabulary items in all of the
+        ``Fields`` in this ``Instance``.
+        """
+        for _, field in self._fields.items():
+            if field.needs_indexing():
+                field.count_vocab_items(counter)
+
+    def index_fields(self, vocab: Vocabulary):
+        """
+        Converts all ``UnindexedFields`` in this ``Instance`` to ``IndexedFields``, given the
+        ``Vocabulary``.  This `mutates` the current object, it does not return a new ``Instance``.
+        """
+        for _, field in self._fields.items():
+            if field.needs_indexing():
+                field.index(vocab)
+
+    def get_padding_lengths(self) -> Dict[str, Dict[str, int]]:
+        """
+        Returns a dictionary of padding lengths, keyed field name.  Each ``Field`` returns a
+        mapping from padding keys to actual lengths, and we just key that dictionary by field name.
+        """
+        lengths = {}
+        for key, field in self._fields.items():
+            lengths[key] = field.get_padding_lengths()
+        return lengths
+
+    def pad(self, padding_lengths: Dict[str, Dict[str, int]]) -> Dict[str, List[numpy.array]]:
+        """
+        Pads each ``Field`` in this instance to the lengths given in ``padding_lengths`` (which is
+        keyed by field name, then by padding key, the same as the return value in
+        :func:`get_padding_lengths`), returning a list of numpy arrays for each field.
+        """
+        arrays = {}
+        for key, field in self._fields.items():
+            field_lengths = padding_lengths.get(key, {})
```

Comment:
In what case would `padding_lengths` not contain the key?  It's always set in `get_padding_lengths`.  If not possible better to fail with an exception here then silently.